plugins/flytekit-pydantic/flytekitplugins/pydantic/object_store.py

```diff
@@ -0,0 +1,107 @@
+import uuid
+from typing import Any, Dict, Type, TypeVar, cast
+
+import pandas as pd
+import torch.nn as nn
+from typing_extensions import Annotated, NewType
+
+from flytekit.core import context_manager, type_engine
+from flytekit.models import literals
+from flytekit.types import directory
+from flytekit.types.file import file
+
+PYDANTIC_SUPPORTED_FLYTE_TYPES = (
+    nn.Module,
+    pd.DataFrame,
+    file.FlyteFile,
+    directory.FlyteDirectory,
+    # TODO - add all supported types
+)
+
+LiteralObjID = Annotated[str, "Key for unique object in literal map."]
+PythonType = TypeVar("PythonType")  # target type of the deserialization
+
+
+class FlyteObjectStore:
+    """
+    This class is an intermediate store for python objects that are being serialized/deserialized.
+
+    On serialization of a basemodel, flyte objects are serialized and stored in this object store.
+    On deserialization of a basemodel, flyte objects are deserialized from this object store.
+    """
+
+    _literal_store: Dict[LiteralObjID, literals.Literal] = {}
+
+    def __setattr__(self, name: str, value: Any) -> None:
+        raise Exception("Attributes should not be set on the FlyteObjectStore.")
+
+    def __init__(self) -> None:
+        raise Exception("This should not be instantiated, it is a singleton object store.")
```

Comment:
Question: what happens when in a single process multiple base models get (de-)serialized? Are we at risk that data serialized for one model is available when deserializing another model? I wonder whether not making this a singleton would be safer. In this case it could have a `to_literal` and `from_literal` method and would be instantiated both in the `to_literal` and the `to_python_value` method of the type transformer.

Reply:
Good question. This was something that took some design on my side. 

I went for the singleton pattern because the validators need to be set on supported flyte objects **before** they are set as a fields in a basemodel subclass. This is because the pydantic.BaseModel gathers all its validators by calling __get_validators__ on each field **at the moment that it is subclassed**, without awareness of which pydantic.BaseModel it is that is calling it.
The singleton pattern where all serialized flyteobjects get saved to the same store enusres that we have all the serialized flyte types in the literalmap that gets sent. While I agree that a single BaseModel independent store is not as slick as having one per item, I went for this because its simplicity. Because we are using UUIDs + class name as the keys in the object store, the risk of two different objects sharing the same key is pretty low. 

Nevertheless, I appreciate your point and I think it could be accommodated without too much changes. We can use a context manager in the to_literal, which creates a namespace that is unique for the current basemodel, locks the access to the object store s.t. all calls to register register_python_object will write into that namespace, and on __exit__ returns the literalmap + releases the lock. Ill commit a proposal for it in a bit.

Reply:
Thanks for this thought. Solved now - think this improved readability of the code easier too!